// errsole/lib/fork/hooks/express/index.js
'use strict'

var appConfig = require('../../../config');

var ExpressHook = {};

function wrapExpress(_express) {
  var express = function createApplication() {
    var app = _express();
    return app;
  };
  Object.setPrototypeOf(express, _express);
  return _express;
}

function wraplisten(_listen, app) {
  return function () {
    if(process.env && process.env.ERRSOLE_MAIN_APP_PORT && process.env.ERRSOLE_APP_PORT && process.env.ERRSOLE_PORTS) {
      arguments[0] = process.env.ERRSOLE_APP_PORT;
    }
    return _listen.call(app, ...arguments);
  }
}

function wrapExpressSession(_session) {
  var MemoryStore = _session.MemoryStore;
  var sessionStore = new MemoryStore();
  appConfig.sessionStore = sessionStore;
  var session = function session(options) {
    options.store = sessionStore;
    return _session(options);
  };
  Object.setPrototypeOf(session, _session);
  return session;
}

ExpressHook.setupHandler = function(app) {
  app.listen = wraplisten(app.listen, app);
}

ExpressHook.packages = [
  {
    name: 'express',
    wrapper: wrapExpress
  },
  {
    name: 'express-session',
    wrapper: wrapExpressSession
  }
];

module.exports = ExpressHook;
